README

              ----========================----
         -== Testoob - Testing Out Of (The) Box ==-
              ----========================----

===========================================================
= Description:
===========================================================

Testoob is an advanced Python unit testing framework that
integrates effortlessly with Python's standard 'unittest'
module.

It's designed and built to be easy to use and extend.

===========================================================
= Links:
===========================================================

Home:        http://code.google.com/p/testoob
Discussion:  http://groups.google.com/group/testoob
Old docs:    http://testoob.sourceforge.net
Source code: http://code.google.com/p/testoob/wiki/SourceControl

===========================================================
= Features:
===========================================================

Start using immediately with the 'testoob' executable, or
replace unittest.main() with testoob.main()

Some advanced options:
- Filtering              (see testoob/extractors.py)
- Reporting              (see testoob/reporting.py)
- Testing facilities     (see testoob/testing.py)
- Running (e.g. threads) (see testoob/running.py)
- And more!

See the full running options by running 'testoob -h', and
check the documentation.

===========================================================
= Installation:
===========================================================

See docs/INSTALL

===========================================================
= Testing
===========================================================

On *nix, use GNU make (usually 'make' or 'gmake') from the project root:

  make test    # test with the default python version
  make testall # test with different python versions

On Windows, run from the project root:

  python tests\alltests.py

===========================================================
= License:
===========================================================

Testoob is distributed under the terms of the Apache
License, Version 2.0.  See docs/COPYING for more
information.

===========================================================
= Credits:
===========================================================

Testoob has been created with the help of:

- Ori Peleg (orispammerpel at users.sourceforge.net)
- Barak Schiller (bspamalotschiller at gmail.com)
- Misha Seltzer (mishadontspamme at seltzer dot com)
- Wolfram Kriesing (ilikespamwolfram at kriesing dot de)
- Ronnie van 't Westeinde <Ronnie.gimmespam.van.t.Westeinde at gmail dot com>

To mail us, remove spam-related words from the user names.

Have fun testing!
The Testoob team.



docs/CHANGELOG
                           -== Testoob Changelog ==-

'Testoob' is an advanced testing framework for Python. It is designed to
integrate effortlessly with existing PyUnit (standard module 'unittest') test
suites.

== Testoob 1.15, 6 October, 2009

Revival release -- wouldn't have happened without Ronnie van 't Westeinde's
contributions and prodding!

The new project home is http://code.google.com/p/testoob

Skipping version 1.14 since it's been used to signify some internal patched
versions.

Enhancements:
 * TESTOOB_WARNING_COLOR env var to change color (Ronnie van 't Westeinde)
 * Progress bar improvements (Ronnie van 't Westeinde)
 * Coverage improvements (Ronnie van 't Westeinde)
 * New option '--time-each-test' (by Ziv, Amir, and Anton), #288
 * No deprecation warnings with Python 2.6
 * Better color support detection in win32 with ctypes
 * Preliminary IronPython and Jython support

Bugfixes:
 * Fixed reporter race condition (only happened on Windows)
 * '--stop-on-fail' no longer stops when first test is skipped (Ronnie van 't Westeinde)
 * '--html' works correctly when test + setup/teardown both fail (Ronnie van 't Westeinde)

Internal:
 * Many tests made simpler and less sensitive (more work to be done)
 * Ctrl-C tests work on Windows
 * Some API changes to testoob.collecting
 * Fixed some tests failing on Windows without ctypes available (py2.4 and earlier)

Contributors:
 * This release is entirely made possible by Ronnie van 't Westeinde

== Testoob 1.13, 3 December, 2006

Enhancements:
 * Python 2.5 support, #281
 * New option --rerun-on-fail (added by Ryan Wilcox), #288
 * --vassert now works with 'testoob.testing' functions, #289

Bugfixes:
 * Install Windows executable helper only on Windows, #279
 * Skipped tests don't count as run tests, #271
 * '--processes' fixed (by Chad Whitacre), #287
 * Missing import in tests fixed (by Lukasz Strzygowski)

Internal:
 * Unit tests for 'testoob.running', #155
 * Using optparse's '--version' hook, #285
 * Testoob is tested by a PyBot, #282 (by Grig Gheorghiu)

Contributors:
 Misha Seltzer, Ori Peleg, Ryan Wilcox, Chad Whitacre, Lukasz Strzygowski,
 Wolfram Kriesing, Grig Gheorghiu

== Testoob 1.12, 9 September, 2006

Enhancements:
 * Added color support on Windows consoles, #260 and #261
 * Added '--timeout-with-threads', which also works on Windows, #262
 * Ctrl-C skips one test, 2 in succession skips to the end, #267

Bugfixes:
 * On Windows, 4Suite is chosen before MSXML so the "date" field in HTML
   output works. #256
 * --pbar fixes: import (#257), race condition (#258), and passing exceptions (#259)

Internal:
 * Better release testing (no path mangling, and added isolated tests)
 * More unit tests, #155
 * Many code cleanups and refactorings
 * Updated release script to use pysvn, #268

Contributors:
 Misha Seltzer, Amir Szekely, Barak Schiller, Michael McFaden, Ori Peleg

== Testoob 1.11, 1 July, 2006

Bugfix release, some options wouldn't run (Thanks Jochen!)

== Testoob 1.1, 24 June, 2006

Enhancements:
 * Optimizations have made Testoob over 4 times faster, #233
 * A summary of failed tests is now printed after the run, #252
 * Added a simple Tk GUI progress report with --pbar, #253
 * A better assertRaises, #42
 * Profiling support with --profile, #163 + #164
 * Silent running with --silent, #225
 * Graceful behavior on interrupt (Ctrl-C), #231
 * Optionally skip a command-line test _after_ the command is run, #241

Bugfixes:
 * An installation bug cause HTML reports to not work, #219
 * Coverage ignores C extensions, #205
 * Coverage now works with Python2.2, #198
 * --timed-repeat calls setUp/tearDown, #201
 * Fixed coverage error due to a bug in std. module 'trace', #206
 * Exit with error on failure also when using testoob.main, #226
 * --repeat and --list now work together, #232
 * Fixed XML import error, #234
 * Skipping tests now works with --processes, #235
 * Fixed html_standalone.py synax error, #247
 * Ignore duplicate suites on the command line, #172
 * Fixed passing arguments with values to testoob.main(), #173

Internal:
 * Added a plugin mechanism for command-line options, #175
 * Cleaned up ErrInfo, TestInfo, and PyroRunner, #236
 * Limit no. of profiling lines printed, #242
 * Show where the profiling stats are saved, #240
 * Check for 'profile' module, not included by default in Debian, #238
 * Proper error message when hotshot stats file is corrupt, #237

New contributors:
 * Aya Horwitz
 * Roi "Kipi" Degani
 * Tomer Alon
 * Yoram Bachrach

Full tracking at https://opensvn.csie.org/traccgi/testoob/trac.cgi/milestone/version1.1

== Testoob 1.0, 28 April, 2006

Many thanks to Nir Soffer and Bryce Hendrix, who both submitted an amazing
amount of bug reports and patches!!!

We also welcome back Barak Schiller, who has returned after an absence.

Enhancements:
 * Automatic color detection
 * Support for skipping tests
 * PDF reports (needs ReportLab's PDF library: http://www.reportlab.org)
 * Improved reporting when testing command-line applications
 * Can pass options through main() (thanks Ryan!)
 * Print current iteration when repeating tests
 * Test-loading glob and regex patterns with --test-method-{glob,regex}
 * Helper functions for loading tests from the local directory,
   (testoob.collecting.collect_from_files)
 * Coverage results reported in xml (thanks Bryce!)

Changes:
 * Twisted is not necessary for running in threads anymore (thanks nirs!)
 * XML file format changed to more closely match that of Ant's JUnitReport

Bugfixes:
 * Zero tests are now correctly reported as an error
 * Coverage now works with Python 2.3
 * Running a nonexistant test gives an error message instead of a traceback
 * Fixed a bug with pickling
 * ReporterProxy now thread safe when threads are used

Internal:
 * Many more unit tests
 * Improved tests structure
 * Much refactoring

Full tracking at https://opensvn.csie.org/traccgi/testoob/trac.cgi/milestone/version1.0

== Testoob 0.9, 17 March, 2006


win/runtests.py
import os, sys, re

script_dir = os.path.dirname(__file__)
base_dir = os.path.abspath(os.path.join(script_dir, ".."))

def die(msg):
    print>>sys.stderr, msg
    sys.exit(1)

def append_to_pathvar(varname, path):
    path_elements = os.environ.get(varname,"").split(os.pathsep)
    os.environ[varname] = os.pathsep.join(path_elements + [path])

append_to_pathvar('PYTHONPATH', os.path.join(base_dir, "src"))
append_to_pathvar('IRONPYTHONPATH', os.path.join(base_dir, "src"))

os.environ["TESTOOB_DEVEL_TEST"] = "1"

class Exec:
    def __init__(self, py_args=None, testoob_extra_args=None):
        self.py_args = py_args or []
        self.testoob_extra_args = testoob_extra_args or []

PYTHON_EXECUTABLES = {
    "2.6" : Exec([r"C:\Python26\python.exe"]),
    "2.5" : Exec([r"C:\Python25\python.exe"]),
    "2.4" : Exec([r"C:\Python24\python.exe"]),
    "2.3" : Exec([r"C:\Python23\python.exe"]),
    "2.2" : Exec([r"C:\Python22\python.exe"], ["--color-mode=never"]), # Python 2.2 pywin32 problems
    "3.1" : Exec([r"C:\Python31\python.exe"]),
    "3.0" : Exec([r"C:\Python30\python.exe"]),
    "2to3" : Exec([r"C:\Python26\python.exe", "-3"]),
    "ipy2.6" : Exec([r"C:\Program Files\IronPython 2.6\ipy.exe"]),
}
DEFAULT_KEY = "2.6"
def get_run_args(args):
    "get_run_args(args) -> py_args(list), testoob_extra_args(list)"
    if len(args) > 0 and args[0] in PYTHON_EXECUTABLES.keys():
        key = args.pop(0) # remove from args list
    else:
        key = DEFAULT_KEY
    if key not in PYTHON_EXECUTABLES:
        die("unknown version '%s', expecting one of %r" % (key, PYTHON_EXECUTABLES.keys()))
    exec_info = PYTHON_EXECUTABLES[key]
    if not os.path.exists(exec_info.py_args[0]):
        die("Can't find python executable '%s'"%exec_info.py_args[0])
    return exec_info.py_args, exec_info.testoob_extra_args + args

print "Can specify python version on the commandline, one of %r" % sorted(PYTHON_EXECUTABLES.keys())
print "Example: '%s 2.6'" % (sys.argv[0])
args = sys.argv[1:]
py_args, testoob_extra_args = get_run_args(args)
testoob_executable = os.path.join(base_dir, "src", "testoob", "testoob")
test_file = os.path.join(base_dir, "tests", "alltests.py")

test_args = py_args + [testoob_executable, test_file, "suite"] + testoob_extra_args
import subprocess
print " ".join(test_args)
subprocess.call(test_args)
